File: JavaScript/1135-Connecting Cities With Minimum Cost.js
/**
 * @param {number} n
 * @param {number[][]} connections
 * @return {number}
 */

// Kruskal's algorithm
var minimumCost = function(n, connections) {
    let parent = new Map();
    for (let i=1; i<=n; i++){
        parent.set(i, i);
    }
    
    connections.sort(function(a,b){
        return a[2]-b[2];
    })
    
    let find = (x) =>{
        if (parent.get(x) === x){
            return x;
        }else{
            parent.set(x, find(parent.get(x)));
            return parent.get(x);
        }
    }
    
    count = 1;
    totCost = 0;
    
    for (let [u, v, cost] of connections){
        if (find(u) !== find(v)){
            count += 1;
            totCost += cost;
            if (count === n) return totCost;
            //union
            parent.set(parent.get(u), parent.get(v));
        }
    }
    
    return -1;
};

File: C++/1135-Connecting Cities With Minimum Cost.cpp
struct UnionFind {
    vector<int> parent, rank;
    explicit UnionFind(int n) : parent(vector<int>(n)), rank(vector<int>(n, 1)) {
        iota(parent.begin(), parent.end(), 0);
    }
    int find(int node) {
        if(node == parent[node]) return node;
        return parent[node] = find(parent[node]);
    }
    bool unite(int node1, int node2) {
        int p1 = find(node1);
        int p2 = find(node2);
        if(p1 == p2) return false; // Same parent
        if(rank[p1] >= rank[p2]) {
            // Append p2 under p1
            parent[p2] = p1;
            rank[p1] += rank[p2];
        }else {
            parent[p1] = p2;
            rank[p2] += rank[p1];
        }
        return true;
    }
};
class Solution {
public:
    int minimumCost(int n, vector<vector<int>>& connections) {
        UnionFind uf(n+1);
        sort(begin(connections), end(connections), [&](auto &a, auto &b) {
            return a[2] < b[2];
        });
        
        int cnt = 0, res = 0;

        for (auto &c:connections) {
            int u = c[0], v = c[1], w = c[2];
            if (uf.find(u)!= uf.find(v)) {
                uf.unite(u, v);
                cnt++;
                res += w;
            }
        }

        return (n-cnt) == 1?res:-1;
    }
};